backend/remote-state/s3/backend_test.go

```diff
@@ -249,6 +249,62 @@ func TestBackendExtraPaths(t *testing.T) {
 	}
 }
 
+func TestKeyEnv(t *testing.T) {
+	testACC(t)
+	bucketName := fmt.Sprintf("terraform-remote-s3-test-%x", time.Now().Unix())
+	keyName0 := "tfstate"
+	keyName1 := "ws1/tfstate"
+	keyName2 := "ws1/env1/tfstate"
+
+	b0 := backend.TestBackendConfig(t, New(), map[string]interface{}{
+		"bucket":               bucketName,
+		"key":                  keyName0,
+		"encrypt":              true,
+		"workspace_key_prefix": "",
+	}).(*Backend)
+
+	b1 := backend.TestBackendConfig(t, New(), map[string]interface{}{
+		"bucket":               bucketName,
+		"key":                  keyName1,
+		"encrypt":              true,
+		"workspace_key_prefix": "root/userA",
+	}).(*Backend)
+
+	b2 := backend.TestBackendConfig(t, New(), map[string]interface{}{
+		"bucket":               bucketName,
+		"key":                  keyName2,
+		"encrypt":              true,
+		"workspace_key_prefix": "root/userA",
+	}).(*Backend)
+
+	if err := testGetWorkspaceForKey(b0, "tfstate", ""); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := testGetWorkspaceForKey(b0, "ws1/tfstate", "ws1"); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := testGetWorkspaceForKey(b1, "root/userA/ws1/tfstate", "ws1"); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := testGetWorkspaceForKey(b1, "root/userA/ws2/tfstate", "ws2"); err != nil {
```

Comment:
Same as above for this check and the one below. The tests are checking the output of `getWorkspaceForKey`, but that doesn't match how it's used by the backend.